client/src/components/HtmlEditorField/HtmlEditorField.js

```diff
@@ -66,7 +66,14 @@ class HtmlEditorField extends TextField {
    * render the editor proper
    */
   renderDependencyScript() {
-    return <Script url={this.props.data.editorjs} onLoad={this.handleReady} />;
+    if (!window.tinymce && !window.TinyMCE) {
```

Comment:
The core issue is that sometimes tinymce is already loaded - whether from another react component (basically another copy of this same component e.g. in a separate elemental area) or from entwine or other logic.

What you've proposed would a) cause issues because you can't set state during a given render cycle, and b) not work, because we're still in a "not ready" state to begin with, which means we'll always end up running `renderDependencyScript()`, which doesn't check if tinymce already exists or not.

I thought about refactoring `render()` itself to be aware of whether a tinymce instance is already loaded, but there is logic in a few places that executes branches of code when the ready state changes - so we _have_ to change the ready state if we detect that tinymce is already loaded.

Calling `handleReady()` in a timeout avoids issues from setting state during the current render cycle, and is reminiscent of how that same function is called after the script has loaded. Basically, we're saying "either load the script, and then call `handleReady()` asynchronously, or if we don't need to load the script, skip straight to calling `handleReady()` asynchronously.

Reply:
OK makes sense

There's some code up the top of the component

```js
    this.state = {
        isReady: !props.data.editorjs
    };
```
where editorjs is the path that later gets mapped to Script.url

Seems like could simply do this? (untested)

```js
    this.state = {
        isReady: !props.data.editorjs || window.TinyMCE || window.tinymce
    };
```
?

Reply:
In theory yes, but that means we don't end up executing the code in `componentDidUpdate()` including `registerChangeListener()` which seems important to run and is the sort of thing that would probably cause a new regression we won't notice for 3 months 😅

Reply:
Can't you just, refactor it as a functional component instead then :-p